Charon/filetypes/GCodeSocket.py

```diff
@@ -0,0 +1,82 @@
+# Copyright (c) 2018 Ultimaker B.V.
+# libCharon is released under the terms of the LGPLv3 or higher.
+#
+# This class is used to read GCode stream that are served
+# dynamically over a TCP connection.
+
+import socket
+import struct
+from io import BytesIO, SEEK_SET, SEEK_CUR
+
+from typing import Any, Dict, IO, Optional, List
+
+from Charon.filetypes.GCodeFile import GCodeFile
+from urllib.parse import urlparse
+
+
+class SocketFileStream(BytesIO):
+    def __init__(self, sock_object: socket.socket):
+        super().__init__()
+        self.current_line = 0
+        self.__socket = sock_object
+
+    def seekable(self) -> bool:
+        return True
+
+    def seek(self, offset: int, whence: Optional[int] = None) -> int:
+        if whence is None or whence == SEEK_SET:
+            self.current_line = offset
+        elif whence == SEEK_CUR:
+            self.current_line += offset
+        else:
+            raise ValueError('Unsupported whence mode in seek: %d' % whence)
+        return offset
+
+    def readline(self, _size: int = -1) -> bytes:
+        self.__socket.send(struct.pack('>I', self.current_line))
+        line = b''
+        char = b''
+
+        while char != b'\n':
+            char = self.__socket.recv(1)
+            line += char
+
+        self.current_line += 1
+        return line
+
+    def read(self, _size: int = -1) -> bytes:
+        assert False
+
+    def readlines(self, _hint: int = -1) -> List[bytes]:
+        assert False
+
+    def tell(self) -> int:
+        assert False
+
+    def close(self):
+        self.__socket.close()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        return self.readline()
+
+
+class GCodeSocket(GCodeFile):
```

Comment:
Our codestyle states that objects should be in seperate files unless they are private functions